src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,154 @@
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(BDX, "failed to handle message: %" CHIP_ERROR_FORMAT, err.Format());
+    }
+
+    // Almost every BDX message will follow up with a response on the exchange. Even messages that might signify the end of a
+    // transfer could necessitate a response if they are received at the wrong time.
+    // For this reason, it is left up to the application logic to call ExchangeContext::Close() when it has determined that the
+    // transfer is finished.
+    ec->WillSendMessage();
+
+    // Get the next output event and send it to the delegate.
+    TransferSession::OutputEvent outEvent;
+    do
+    {
+        mTransfer.GetNextAction(outEvent);
+
+        if (outEvent.EventType == TransferSession::OutputEventType::kTransferTimeout)
+        {
+            OnResponseTimeout(ec);
+        }
+        else
+        {
+            HandleAsyncTransferSessionOutput(outEvent);
+        }
+    } while (outEvent.EventType != TransferSession::OutputEventType::kNone);
+
+    return err;
+}
+
+Messaging::ExchangeContext * AsyncTransferFacilitator::GetExchangeContext()
+{
+    if (mExchange)
+    {
+        return mExchange.Get();
+    }
+    return nullptr;
+}
+
+void AsyncTransferFacilitator::OnExchangeClosing(chip::Messaging::ExchangeContext * ec)
+{
+    ChipLogDetail(BDX, "%s, ec: " ChipLogFormatExchange, __FUNCTION__, ChipLogValueExchange(ec));
+    if (mExchange)
+    {
+        mExchange.Release();
+    }
+}
+
+void AsyncTransferFacilitator::OnResponseTimeout(Messaging::ExchangeContext * ec)
+{
+    ChipLogDetail(BDX, "%s, ec: " ChipLogFormatExchange, __FUNCTION__, ChipLogValueExchange(ec));
+    if (mExchange)
+    {
+        mExchange.Release();
+    }
+}
+
+CHIP_ERROR AsyncResponder::PrepareForTransfer(Messaging::ExchangeContext * exchangeCtx, TransferRole role,
+                                              BitFlags<TransferControlFlags> xferControlOpts, uint16_t maxBlockSize,
+                                              System::Clock::Timeout timeout)
+{
+    VerifyOrReturnError(exchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    ReturnErrorOnFailure(mTransfer.WaitForTransfer(role, xferControlOpts, maxBlockSize, timeout));
+
+    VerifyOrReturnError(!mExchange, CHIP_ERROR_INCORRECT_STATE);
+    mExchange.Grab(exchangeCtx);
+
+    return CHIP_NO_ERROR;
+}
+
+void AsyncResponder::ResetTransfer()
+{
+    if (mExchange)
+    {
+        mExchange.Release();
+    }
+    mTransfer.Reset();
+}
+
+bdx::StatusCode AsyncResponder::GetBdxStatusCodeFromChipError(CHIP_ERROR err)
+{
+    if (err == CHIP_ERROR_INCORRECT_STATE)
+    {
+        return bdx::StatusCode::kUnexpectedMessage;
+    }
+    if (err == CHIP_ERROR_INVALID_ARGUMENT)
+    {
+        return bdx::StatusCode::kBadMessageContents;
+    }
+    return bdx::StatusCode::kUnknown;
+}
+
+void AsyncResponder::NotifyEventHandledWithError(CHIP_ERROR error)
+{
+    ChipLogDetail(BDX, "%s : error %d", __FUNCTION__, error.AsInteger());
```

Comment:
Please no `__FUNCTION__`, and use `error.Format()` for logging errors.

Reply:
done.